### PR TITLE
Drop support for DBAL 2, PHP 7.2, Symfony 3, Persistence 1

### PR DESCRIPTION
**Pull request by greg0ire**
DBAL 2 will no longer be supported tomorrow, and DBAL 3 requires PHP 7.3

### CODE REVIEW

File: lib/Doctrine/Migrations/Tools/TransactionHelper.php [hunk 1]
```diff
@@ -72,11 +72,9 @@ private static function inTransaction(Connection $connection): bool
...
-            try {
-                return $connection->getNativeConnection();
-            } catch (LogicException $e) {
-            }
+        try {
+            return $connection->getNativeConnection();
+        } catch (LogicException $e) {
         }
 
         $innermostConnection = $connection;
```
Comment (greg0ire) [inline lib/Doctrine/Migrations/Tools/TransactionHelper.php:80]: Todo: check if the code below is still useful

File: .appveyor.yml [hunk 2]
```diff
@@ -13,7 +13,6 @@ cache:
 
 environment:
   matrix:
-    - php: 7.2.25
```
Comment (greg0ire) [inline .appveyor.yml:16]: Todo: determine what our policy for testing on appveyor shall be

File: lib/Doctrine/Migrations/Tools/TransactionHelper.php (hunk 1 above)
Comment (greg0ire) [reply lib/Doctrine/Migrations/Tools/TransactionHelper.php:80]: From what I understood, it will be useful until DBAL 4, when we are sure that `getNativeConnection` can no longer throw a `LogicException`.

**Comment by greg0ire** [comment]
:thinking: just noticed we apparently still support unmaintained versions of Symfony, we should bump that as well.

File: composer.json [hunk 3]
```diff
@@ -24,32 +24,32 @@
         }
     ],
     "require": {
-        "php": "^7.2 || ^8.0",
+        "php": "^7.3 || ^8.0",
```
Comment (derrabus) [inline composer.json]: Shall we drop 7.3 as well?

File: composer.json [hunk 4]
```diff
@@ -24,32 +24,32 @@
...
         "composer-runtime-api": "^2",
-        "doctrine/dbal": "^2.11 || ^3.0",
+        "doctrine/dbal": "^3.0",
```
Comment (derrabus) [inline composer.json]: We need to require `^3.3` here because you've dropped some compat code that we would still need for DBAL 3.2 and below, like checking for `getNativeConnection()`.

File: composer.json [hunk 5]
```diff
@@ -24,32 +24,32 @@
...
         "phpstan/phpstan": "^1.5",
         "phpstan/phpstan-deprecation-rules": "^1",
         "phpstan/phpstan-phpunit": "^1.1",
         "phpstan/phpstan-strict-rules": "^1.1",
         "phpstan/phpstan-symfony": "^1.1",
-        "phpunit/phpunit": "^8.5 || ^9.4",
-        "symfony/cache": "^3.4.26 || ^4.2.12 || ^5.0 || ^6.0",
-        "symfony/process": "^3.4 || ^4.0 || ^5.0 || ^6.0",
-        "symfony/yaml": "^3.4 || ^4.0 || ^5.0 || ^6.0"
+        "phpunit/phpunit": "^9.4",
```
Comment (derrabus) [inline composer.json:49]:
```suggestion
        "phpunit/phpunit": "^9.5",
```

PHPUnit 9.4 is dead and buried.

**Review by derrabus — CHANGES_REQUESTED** [review]

File: composer.json (hunk 3 above)
Comment (greg0ire) [reply composer.json]: I don't see a reason to, but also, I don't see a reason not to, so sure, let's do it!

File: phpcs.xml.dist [hunk 6]
```diff
@@ -17,6 +17,18 @@
...
+        <exclude-pattern>*/lib/*</exclude-pattern>
+    </rule>
+
+    <rule ref="SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingNativeTypeHint">
+        <exclude-pattern>*/lib/*</exclude-pattern>
+    </rule>
+
+    <rule ref="SlevomatCodingStandard.TypeHints.PropertyTypeHint.MissingNativeTypeHint">
+        <exclude-pattern>*/lib/*</exclude-pattern>
+    </rule>
```
Comment (greg0ire) [inline phpcs.xml.dist:22]: Maybe I should remove those, `git add --patch` the parts that are backwards compatible, and restore these lines. But that would probably be best in a separate PR for the sake of review.

Comment (greg0ire) [reply phpcs.xml.dist:22]: Just tried it, and it's not that complicated to untangle, I'll push it as an extra commit.

**Comment by greg0ire** [comment]
I think that the 2 issues reported by the BC check can be safely ignored.

**Review by derrabus — APPROVED** [review]
